---
title: "Gene Annotation"
author: "Fatemeh"
date: "June 17, 2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## FinalGeneSetAnnotation
This script has the fallowing functions:\
\
1. annotate.final.geneset()\
This will take integrated_tse_ara.txt as input, filters genes based on sme criteria and prepares the final genes set.\
\
```{r , echo=FALSE,include=FALSE}
  library(gsubfn)
  library(ggplot2)
  library(seqinr)
  library(gdata)
  library(Biostrings)
  library(plyr)
  library(ggplot2)
  library(gridExtra)

annotate.final.geneset <- function() {

  genefilepath <-
  "/home/fatemeh/TriTrypGenome_project/GenePrediction/integrated_tse_ara.txt"
  genefile <-
  read.table(genefilepath, header = TRUE, colClasses = "character")
  
  # cuttoff score for TSE 50, for ARA 107
  genefile[genefile$arascore == "notfound",]$arascore <- -1
  genefile$arascore <- as.integer(genefile$arascore)
  genefile[genefile$tsescore == "notfound",]$tsescore <- -1
  genefile$tsescore <- as.integer(genefile$tsescore)
  
  dismiss0 <-
  (genefile$foundby == "both") &
  (genefile$arascore < 107 | genefile$tsescore < 50)
  dismiss1 <-
  (genefile$foundby == "ara") &
  (genefile$arascore < 107) # most of these genes were pseudo or truncated or both
  dismiss2 <- (genefile$foundby == "tse") & (genefile$tsescore < 50)
  genefile <- genefile[(!dismiss0 & !dismiss1 & !dismiss2),]
  
  # genefunc column is added which shows the presented gene identity in table
  # Genes marked as ?? include: pseudo|truncated genes, genes with unmatched identity, genes with unassigned identity|anticodon by any of genefinders.
  
  ambiguty1 <-
  genefile$tsefunc == "" &  genefile$arafunc == "" # 2 genes
  ambiguty2 <-
  genefile$tsenote != "notfound" # 2 genes not the same 2 genes in ambiguty1
  ambiguty3 <-
  genefile$foundby == "both" &
  genefile$tsefunc != genefile$arafunc # 22 genes
  ambiguty4 <- logical(length = nrow(genefile))
  for (i in 1:nrow(genefile)) {
  if (genefile$foundby[i] != "ara")
  ambiguty4[i] <-
  length(grep("n|N", genefile$tsegeneseq[i])) == 1
  else
  ambiguty4[i] <-
  length(grep("n|N", genefile$arageneseq[i])) == 1
  }
  
  ambiguties <- ambiguty1 | ambiguty2 | ambiguty3 | ambiguty4
  genefile$genefunc <- ""
  genefile[ambiguties,]$genefunc <- "??"
  
  #table(genefile[!ambiguties, ]$foundby)
  #ara both
  #36 3525
  
  genefile[!ambiguties,]$genefunc <-
  genefile[!ambiguties,]$arafunc
  
  # resctoring 10 genes with ambiguties Y|N
  # they were genes with exact same sequence. with relatively high TSE score of 72 and relatively low ARA score of 107.
  # 9 of these genes are from clade Tcruzi and complete 8 cruzi genomes which do not have any Y genes annotated for them in figure.
  genefile[genefile$genefunc == "??" &
  genefile$tsefunc == "Y", ]$genefunc <- "Y"
  genefile
}
genefile <- annotate.final.geneset()
```
\
2. Score.visualization()\
This functions shows the distribution of gene scores \
```{r plot, echo=FALSE}

geneDF <- read.table("/home/fatemeh/TriTrypGenome_project/GenePrediction/integrated_tse_ara.txt",header = TRUE)
geneDF2 <- geneDF[geneDF$foundby == "ara" | geneDF$foundby == "both",]
geneDF2$arascore <- as.double(geneDF2$arascore)
geneDF2$Foundby <- "Aragorn-only"
geneDF2[geneDF2$foundby!="ara",]$Foundby <- "Both"
geneDF2[geneDF2$tsenote!="notfound",]$Foundby <- "Pseudo/Truncated"
geneDF2$Foundby <- factor(geneDF2$Foundby,levels=c("Both","Aragorn-only","Pseudo/Truncated"))
#"Aragorn score of genes found by both genefinders TSE and ARA, and genes found by only ARA"
araplot <- ggplot(geneDF2,
         aes(
           x = arascore,
           fill = Foundby
         )) + theme_bw() +
    geom_density(alpha=0.3) + xlab("Aragorn Score") + ggtitle("Aragorn Score Distributions") +    theme(legend.title=element_blank())

geneDF3 <- geneDF[geneDF$foundby == "tse" | geneDF$foundby == "both",]
geneDF3$tsescore <- as.double(geneDF3$tsescore)
geneDF3$Foundby <- "tRNAscan-SE2-only"
geneDF3[geneDF3$foundby =="both",]$Foundby <- "Both Gene-Finders"
geneDF2$Foundby <- factor(geneDF2$Foundby,levels=c("Both Gene-Finders","tRNAscan-SE2-only"))
tseplot <- ggplot(geneDF3,
         aes(
           x = tsescore,
           fill = Foundby
         )) + theme_bw() +
    geom_density(alpha=0.3) + xlab("tRNAscan-SE2 Score") + ggtitle("tRNAscan-SE2 Score Distributions") + theme(legend.title=element_blank())
grid.arrange(araplot, tseplot, ncol=1)
```
\newpage
```{r summary, echo=FALSE}
gcContent <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("G", "C")]) * 100
}
atContent <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("A", "T")]) * 100
}
Apercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("A")]) * 100
}
Cpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("C")]) * 100
}
Gpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("G")]) * 100
}
Tpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("T")]) * 100
}
```
2. create.summary.table()\
This creates a summary table of selected genes.\
```{r , echo=FALSE}
#create.summary.table <- function(genefile) {
  # create a table with columns: Annotation, Intersection, ARAonly, Union
  firstcol <- c("#tRNA","#N/#G","Min Gene Length","Max Gene Length","%intron","%G" ,"%C" ,"%T" ,"%A","A",  "C",  "D",  "E",  "F",  "G",  "H",  "I",  "K",  "L",  "M",  "N",  "P",  "Q",  "R",  "S",  "T",  "V",  "W",  "X",  "Y",  "Z", "??")
  resulttable <- data.frame(firstcol,rep(0,length(firstcol)),rep(0,length(firstcol)),rep(0,length(firstcol)))
  names(resulttable) <- c("Annotation", "Intersection", "ARAonly", "Union")
  
  isboth <- genefile$foundby=="both"
  bothdf <- genefile[isboth,]
  isaraonly <- genefile$foundby=="ara"
  intersectDF <- genefile[isboth,]
  ARAonlyDF <- genefile[isaraonly,]
  genefile$geneseq <- ""
  genefile[isaraonly,]$geneseq <- genefile[isaraonly,]$arageneseq
  genefile[!isaraonly,]$geneseq <- genefile[!isaraonly,]$tsegeneseq
  
  resulttable[1,2:4] <- c(nrow(intersectDF),nrow(ARAonlyDF),nrow(genefile))
  resulttable[2,2:4] <- c(sum(nchar(intersectDF$tsegeneseq))/nrow(intersectDF),sum(nchar(ARAonlyDF$arageneseq))/nrow(ARAonlyDF),sum(nchar(genefile$geneseq))/nrow(genefile)) # sum of gene length / # genes from the first row
  resulttable[3,2:4] <- c(min(nchar(intersectDF$tsegeneseq)),min(nchar(ARAonlyDF$arageneseq)),min(nchar(genefile$geneseq)))
  resulttable[4,2:4] <- c(max(nchar(intersectDF$tsegeneseq)),max(nchar(ARAonlyDF$arageneseq)),max(nchar(genefile$geneseq)))
  resulttable[5, 2:4] <-
    c(
      100 * nrow(bothdf[bothdf$tseintronbegin != 0, ]) / nrow(bothdf),
      100 * nrow(ARAonlyDF[ARAonlyDF$araintronbegin != "nointron", ]) / nrow(ARAonlyDF),
      100 * (nrow(bothdf[bothdf$tseintronbegin != 0, ]) + nrow(ARAonlyDF[ARAonlyDF$araintronbegin !=
                                                                           "nointron", ])) / nrow(genefile)
    )
  resulttable[6,2:4] <- c(Gpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Gpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Gpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[7,2:4] <- c(Cpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Cpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Cpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[8,2:4] <- c(Tpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Tpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Tpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[9,2:4] <- c(Apercentage(paste(intersectDF$tsegeneseq,collapse = "")),Apercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Apercentage(paste(genefile$geneseq,collapse = "")))
  
  # make a table of Class frequencies for each set:
  intersect_ClassFreq <- as.data.frame(table(intersectDF$genefunc))
  names(intersect_ClassFreq) <- c("class","freq")
  AraOnly_ClassFreq <- as.data.frame(table(ARAonlyDF$genefunc))
  names(AraOnly_ClassFreq) <- c("class","freq")
  Union_ClassFreq <- as.data.frame(table(genefile$genefunc))
  names(Union_ClassFreq) <- c("class","freq")
  intersect_ClassFreq$class <- as.character(intersect_ClassFreq$class)
  AraOnly_ClassFreq$class <- as.character(AraOnly_ClassFreq$class)
  Union_ClassFreq$class <- as.character(Union_ClassFreq$class)
  
  for (i in 1:nrow(intersect_ClassFreq)) {
    curr_class <- intersect_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,2] <- intersect_ClassFreq$freq[i]
  }
  for (i in 1:nrow(AraOnly_ClassFreq)) {
    curr_class <- AraOnly_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,3] <- AraOnly_ClassFreq$freq[i]
  }
  for (i in 1:nrow(Union_ClassFreq)) {
    curr_class <- Union_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,4] <- Union_ClassFreq$freq[i]
  }
  
  resulttable$Intersection <- round(resulttable$Intersection,digits = 0)
  resulttable$ARAonly <- round(resulttable$ARAonly,digits = 0)
  resulttable$Union <- round(resulttable$Union,digits = 0)
  resulttable$Annotation <- as.character(resulttable$Annotation)
  #Latex.file.prep(resulttable,"resulttable")
  resulttable
#}
```
\newpage
3. clustersize.dist.visualize()\
This function visualizes the Cluster size distribution for three categories of TryTryp genomes. Labels in green on top of each bar show the percentage of total number of genes as cluster of a specific length. Each color refers to one category of TriTryp genomes. Numbers within each color section of the bar shows the counts of clusters with a specific length.\
```{r ,echo=FALSE, include=FALSE}
library(gsubfn)
library(ggplot2)
library(seqinr)
library(gdata)
library(ggplot2)
library(reshape2)
library(plyr)
#source("https://bioconductor.org/biocLite.R")
library(Biostrings)
create.clusterDF <- function(tsfmGeneDF){
  geneset <- assignCluster(tsfmGeneDF,1000)
  m <- matrix(nrow = max(geneset$cluster),ncol = 9)
  clusterDF<- as.data.frame(m)
  names(clusterDF) <- c("sourceOrg","soureSeq","clusterID","clusterSeq1","clusterSeq2", "clusterBegin","clusterEnd","clusterDir","clusterLen")
  for (i in 1:max(geneset$cluster)) {
    clusterDF$sourceOrg[i] <- geneset[geneset$cluster==i,]$sourceOrg[1]
    clusterDF$soureSeq[i] <- geneset[geneset$cluster==i,]$sourceseq[1]
    clusterDF$clusterID[i] <- i
    clusterDF$clusterSeq1[i] <- paste(geneset[geneset$cluster==i,]$genecode,collapse = ",")
    clusterDF$clusterSeq2[i] <- paste(geneset[geneset$cluster==i,]$genefunc,collapse = "")
    clusterDF$clusterBegin[i] <- sort(geneset[geneset$cluster==i,]$begin)[1]
    clusterDF$clusterEnd[i] <- sort(geneset[geneset$cluster==i,]$end)[nrow(geneset[geneset$cluster==i,])]
    clusterDF$clusterDir[i] <- paste(geneset[geneset$cluster==i,]$direction,collapse = "")
    clusterDF$clusterLen[i] <- nrow(geneset[geneset$cluster==i,])
  }
  clusterDF$class <- toupper(substr(clusterDF$sourceOrg,1,1))
  clusterDF[clusterDF$class!="T" & clusterDF$class!="L",]$class <- "O"
  clusterDF
}
assignCluster <- function(geneDF, clusterdistance) {
  
  geneDF$tsebegin <- as.integer(geneDF$tsebegin)
  geneDF$tseend <- as.integer(geneDF$tseend)
  geneDF$arabegin <- as.integer(geneDF$arabegin)
  geneDF$araend <- as.integer(geneDF$araend)
  
  geneDF$begin <- geneDF$tsebegin
  geneDF[geneDF$foundby=="ara",]$begin <- geneDF[geneDF$foundby=="ara",]$arabegin
  geneDF$end <- geneDF$tseend
  geneDF[geneDF$foundby=="ara",]$end <- geneDF[geneDF$foundby=="ara",]$araend
  
  geneDF <-
    geneDF[order(geneDF$sourceOrg, geneDF$sourceseq, geneDF$begin), ]
  geneDF$cluster <- 0
  geneDF$cluster[1] <- 1
  setnumber <- 1
  
  for (i in 1:(nrow(geneDF) - 1)) {
    geneDist <-
      abs(geneDF$begin[i + 1] - geneDF$begin[i])
    if ((geneDist < clusterdistance) &&
        (geneDF$sourceOrg[i] == geneDF$sourceOrg[i + 1]) &&
        (geneDF$sourceseq[i + 1] == geneDF$sourceseq[i]))
      geneDF$cluster[i + 1] <- setnumber
    else
    {
      setnumber <- setnumber + 1
      geneDF$cluster[i + 1] <- setnumber
    }
  }
  geneDF
}
```
```{r ,echo=FALSE, warning=FALSE}
#clustersize.dist.visualize <- function(genefile){
  genefile <- annotate.final.geneset()
  clusterdf <- create.clusterDF(genefile)
  # dataframe c("clusterLen","Tfreq","Lfreq","Ofreq","percOfgenes")
  maxcluslen <- max(nchar(clusterdf$clusterSeq2))
  sizeDF <- data.frame(seq(1,maxcluslen,1),rep(0,maxcluslen))
  names(sizeDF) <- c("ClusterLen","Tfreq")
  sizeDF$Lfreq <- 0
  sizeDF$Ofreq <- 0
  sizeDF$percOfgenes <- 0
  for (i in 1: nrow(sizeDF)) {
    sizeDF$Tfreq[i] <- nrow(clusterdf[clusterdf$class=="T" & nchar(clusterdf$clusterSeq2) == i,])
    sizeDF$Lfreq[i] <- nrow(clusterdf[clusterdf$class=="L" & nchar(clusterdf$clusterSeq2) == i,])
    sizeDF$Ofreq[i] <- nrow(clusterdf[clusterdf$class=="O" & nchar(clusterdf$clusterSeq2) == i,])
    size_i_freq <- sizeDF$Tfreq[i] + sizeDF$Lfreq[i] + sizeDF$Ofreq[i]
    gene_num <- size_i_freq * i 
    sizeDF$percOfgenes[i] <- round((gene_num/nrow(genefile))*100,digits = 0)
  }
  
  sizeDF2 <- melt(sizeDF, id = c("ClusterLen","percOfgenes"))
  # Sort the data by ClusterLen and variable
  # Calculate the cumulative sum of the variable value for each cluster
  # Create the plot
  sizeDF2$ClusterLen <- factor(sizeDF2$ClusterLen)
  sizeDF2$OLTorder <- 0
  sizeDF2[sizeDF2$variable == "Ofreq", ]$OLTorder <- 0
  sizeDF2[sizeDF2$variable == "Lfreq", ]$OLTorder <- 1
  sizeDF2[sizeDF2$variable == "Tfreq", ]$OLTorder <- 2
  sizeDF2_sorted <-
    sizeDF2[order(sizeDF2$ClusterLen, sizeDF2$OLTorder), ]
  df_cumsum <- ddply(sizeDF2_sorted, "ClusterLen",
                     transform, label_ypos = cumsum(value))
  df_cumsum$label_ypos2 <- ""
  df_cumsum$label2 <- ""
  for (i in 1:nrow(df_cumsum)) {
    if (i %% 3 == 0)
    {
      df_cumsum$label_ypos2[i] <- toString(df_cumsum$label_ypos[i])
      df_cumsum$label2[i] <-
        paste(toString(df_cumsum$percOfgenes[i]), "%", sep = "")
    }
    else
    {
      df_cumsum$label_ypos2[i] = ""
      df_cumsum$label2[i] = ""
    }
  }
  df_cumsum$label1 <- ""
  df_cumsum$label1 <- df_cumsum$value
  df_cumsum[df_cumsum$value < 15,]$label1 <- ""
  ggplot(data = df_cumsum, aes(x = ClusterLen, y = value, fill = variable)) +
    geom_bar(stat = "identity") +
    geom_text(
      aes(y = label_ypos, label = label1),
      vjust = 1.3,
      color = "gray20",
      size = 4
    ) +
    scale_fill_brewer(palette = "Paired") + geom_text(aes(y = as.integer(label_ypos2), label =
                                                            label2),
                                                      vjust = -0.3,
                                                      size = 5,color = "darkgreen") + 
    scale_fill_manual(values = c("#00AFBB", "#E7B800", "lightgray"),name = "Genomes",labels = c("Trypanosoma", "Leishmania", "Others")) +
    theme_minimal() + xlab("Cluster Length") + ylab("Frequency")
  
#}

```
\newpage

\
3. prepare.tsfm.input()\
Writing the seleted genes in file tsfm_input_geneset.txt.\

```{r pressure, echo=FALSE, include=FALSE}
tsfmInput <- genefile[genefile$genefunc!="??" & genefile$genefunc != "Z",]
resultpath <- "/home/fatemeh/TriTrypGenome_project/GeneAnnotation/"
write.table(tsfmInput,col.names = TRUE,file = paste(resultpath,"tsfm_input_geneset.txt",sep = ""))
```
\
The gene set tsfm_input_geneset.txt is passed to the script TriTrypAlignment.R to be aligned with Human tRNA genes.\
\
\

*TriTrypAlignment.R*

Alignment Steps:\
1. Genes from tsfm_input_geneset.txt are read and functional classes are added at the end of the geneID\
2. Variable arms are removed based on reported secondary structure from genefinders\
3. Gene introns are removed based in the secondary sctructure\
4. the result is merged with the Human tRNA genes (the headers for Homo genes are also updated) and the result is saved in file coveainput.fasta.\
5. coveainput.fasta is aligned to the Eukaryote model using covea\
6. the result (Aligned_TriTryp_Homo.covea) is edited based on the fallowing criteria in order:\
   a)  sites that have more than 98% gap are removed\
   b)  sequences with more than 3 gaps are removed\
   c)  sequences with two or more gaps next to eachother are removed\
7. the alignment result is saved as fasta file in Aligned_TriTryp_Homo.fasta, with secondary structure saved as Aligned_TriTryp_Homo_structfile.txt\
\
\
*SplitAlignedGenes.R*
\
The alignment Result Aligned_TriTryp_Homo.fasta will be passed to this script to be splitted either by genome, or clusters of genomes.\
The result fasta file for each genome is saved as a file in tsfm/input folder.\
The missing functional class for each genome or cluster of genomes is visualized as a bar plot.\
\
```{r , echo=FALSE, include=FALSE}
library(ggplot2)
library(gdata)
#split_tRNAgene_with_clustername <- function() {
  resultpath <- "/home/fatemeh/TriTrypGenome_project/tsfm/input2/"
  fastafile <-
    read.table(
      "/home/fatemeh/TriTrypGenome_project/GeneAnnotation/Aligned_TriTryp_Homo.fasta",
      sep = "\n"
    )
  fastafile <- as.character(fastafile$V1)
  sequences <-
    as.character(fastafile[seq(2, length(fastafile), 2)])
  headers <- as.character(fastafile[seq(1, length(fastafile), 2)])
  headers <- gsub("_ara", "_", headers)
  headers <- gsub("\\s", "", headers)
  
  headers2 <-
    lapply(X = headers , function(X)
      gsub("Homo", "HOMO_", X, fixed = TRUE))
  headers2 <- as.character(headers2)
  
  genomeName <-
    lapply(X = headers2 , function(X)
      gsub(">", "", X, fixed = TRUE))
  genomeName <- as.character(genomeName)
  
  genomeName2 <-
    lapply(X = genomeName, function(X)
      unlist(strsplit(X, split = "_"))[1])
  genomeName2 <- as.character(genomeName2)
  
  # we will split genomes based on clusters of genomes
  # add another column "clustername" to tRNAdf to assign the new names to each cluster of genomes:
  clusNames <- character(length = length(headers))
  clusNames <- genomeName2
  for (i in 1:length(clusNames)) {
    clusternames <- clusNames[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      clusNames[i] <- "LenriettiComplex"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      clusNames[i] <- "AfricanTrypanosome"
    #clusternames == "TcruziCLBrener" clusternames == "TrangeliSC58" removed
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      clusNames[i] <- "AmericanTrypanosome"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      clusNames[i] <- "Leishmania1"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      clusNames[i] <- "Leishmania3"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      clusNames[i] <- "Leishmania4"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      clusNames[i] <- "Leishmania2"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      clusNames[i] <- "Leishmania5"
  }
  
  functionalclasses <-
    lapply(X = headers, function(X)
      unlist(strsplit(X, split = "_"))[length(unlist(strsplit(X, split = "_")))])
  
  functionalclasses <- as.character(functionalclasses)
  
  tRNAdf <-
    data.frame(genomeName2,
               sequences,
               headers2,
               functionalclasses,
               clusNames)
  names(tRNAdf) <-
    c("genomename", "sequence", "headers", "funclass", "clusNames")
  
  genome_list = split(tRNAdf, f = tRNAdf$clusNames)
  
  # make a file for each genome and put the genomes in there
  for (i in 1:length(genome_list)) {
    # write the DF in a file to be processed by the bash
    filepath <-
      paste(resultpath,
            names(genome_list[i]),
            ".fasta",
            sep = "")
    
    write.fwf(
      data.frame(genome_list[[i]]$headers,
                 genome_list[[i]]$sequence),
      filepath,
      sep = "\n",
      colnames = FALSE
    )
  }
 # vidualization(genome_list, )
#}
```

```{r , echo=FALSE, include=FALSE}
#s  resultpath <- "/home/fatemeh/TriTrypGenome_project/tsfm/input1/"
  fastafile <-
    read.table(
      "/home/fatemeh/TriTrypGenome_project/GeneAnnotation/Aligned_TriTryp_Homo.fasta",
      sep = "\n"
    )
  fastafile <- as.character(fastafile$V1)
  sequences <-
    as.character(fastafile[seq(2, length(fastafile), 2)])
  headers <- as.character(fastafile[seq(1, length(fastafile), 2)])
  headers <- gsub("_ara", "_", headers)
  headers <- gsub("\\s", "", headers)
  
  headers2 <-
    lapply(X = headers , function(X)
      gsub("Homo", "HOMO_", X, fixed = TRUE))
  headers2 <- as.character(headers2)
  
  genomeName <-
    lapply(X = headers2 , function(X)
      gsub(">", "", X, fixed = TRUE))
  genomeName <- as.character(genomeName)
  
  genomeName2 <-
    lapply(X = genomeName, function(X)
      unlist(strsplit(X, split = "_"))[1])
  genomeName2 <- as.character(genomeName2)
  
  
  functionalclasses <-
    lapply(X = headers, function(X)
      unlist(strsplit(X, split = "_"))[length(unlist(strsplit(X, split = "_")))])
  
  functionalclasses <- as.character(functionalclasses)
  
  tRNAdf <-
    data.frame(genomeName2, sequences, headers2, functionalclasses)
  names(tRNAdf) <-
    c("genomename", "sequence", "headers", "funclass")
  
  genome_list = split(tRNAdf, f = tRNAdf$genomename)
  
  # make a file for each genome and put the genomes in there
  for (i in 1:length(genome_list)) {
    # write the DF in a file to be processed by the bash
    filepath <-
      paste(resultpath,
            names(genome_list[i]),
            ".fasta",
            sep = "")
    
    write.fwf(
      data.frame(genome_list[[i]]$headers,
                 genome_list[[i]]$sequence),
      filepath,
      sep = "\n",
      colnames = FALSE
    )
    
  }
#  vidualization(genome_list, resultpath)
#}
```

```{r , echo=FALSE}
  tRNAcountdf <- data.frame(names(genome_list))
  tRNAcountdf$counts <- 0
  names(tRNAcountdf) <- c("genome", "tRNAcounts")
  for (i in 1:length(genome_list)) {
    tRNAcountdf[tolower(tRNAcountdf$genome) == tolower(names(genome_list[i])),]$tRNAcounts <-
      nrow(genome_list[[i]])
  }
  
  tRNAcountdf <- tRNAcountdf[order(tRNAcountdf$tRNAcounts), ]
  tRNAcountdf$genome <-
    factor(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  tRNAcountdf$genome <-
    as.character(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  
  func_classes <-
    c(
      "A",
      "R",
      "N" ,
      "D",
      "C" ,
      "Q",
      "E",
      "G",
      "H",
      "I",
      "L",
      "K",
      "M" ,
      "X",
      "F",
      "P",
      "S",
      "T",
      "W",
      "Y" ,
      "V"
    )
  tRNAcountdf$percent_21 <- 0
  tRNAcountdf$missing <- ''
  for (i in 1:length(genome_list)) {
    if (tolower(names(genome_list[i])) != "homo")
    {
      countsdf <-
        as.data.frame(table(as.character(genome_list[[i]]$funclass)))
      tRNAcountdf[tRNAcountdf$genome == names(genome_list[i]), ]$missing <-
        paste(setdiff(func_classes, as.character(countsdf$Var1)), collapse = " ")
      diffset <-
        setdiff(func_classes, as.character(countsdf$Var1))
      fcount <- 21 - length(diffset)
      tRNAcountdf[tRNAcountdf$genome == names(genome_list[i]), ]$percent_21 <-
        (fcount / 21) * 100
    }
    else
      tRNAcountdf[tolower(tRNAcountdf$genome) == tolower(names(genome_list[i])), ]$percent_21 <-
        100
    
  }
  tRNAcountdf$genome <-
    factor(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  p <- ggplot(data = tRNAcountdf, aes(x = genome, y = percent_21)) +
    geom_bar(stat = "identity", fill = "#56B4E9")  + theme(
      axis.text.x = element_text(angle = 90, hjust = 1),
      plot.title = element_text(hjust = 0.5) ,
      text = element_text(size = 8)
    ) +
    geom_text(
      aes(y = tRNAcountdf$percent_21, label = tRNAcountdf$tRNAcounts),
      vjust = 1.3,
      color = "gray20",
      size = 2
    ) +
    labs(title =
           "Percentage of 22 tRNA functional classes covered by each cluster",
         x =
           "Genome", y = "Percentage of tRNA functional classes(total=22)") + geom_text(
             aes(label =
                   missing),
             vjust = 0.5,
             angle = 90,
             hjust = -0.1,
             color = "blueviolet",
             size = 2
           )
  p
  # ggsave(
  #   paste(plotpath, "funcPerc_clustered2.png", sep = ""),
  #   width = 14,
  #   height = 9
  # )
```
\
\
\
Fasta gene files will be splitted based on tRNA functional class by running script splitFuncClass.sh.\

